wrapper/src/test/java/integration/util/AuroraTestUtility.java

```diff
@@ -344,14 +430,342 @@ public void deleteCluster() {
+      throw new RuntimeException(
+          "Cannot find cluster member whose db instance identifier is " + instanceId);
+    }
+    return matchedMemberList.get(0);
+  }
+
+  public List<DBClusterMember> getDBClusterMemberList(String clusterId) {
+    final DBCluster dbCluster = getDBCluster(clusterId);
+    return dbCluster.dbClusterMembers();
+  }
+
+  public void makeSureInstancesUp(List<String> instances) throws InterruptedException {
+    makeSureInstancesUp(instances, true);
+  }
+
+  protected void makeSureInstancesUp(List<String> instances, boolean finalCheck)
+      throws InterruptedException {
+    final ExecutorService executorService = Executors.newFixedThreadPool(instances.size());
+    final ConcurrentHashMap<String, Boolean> remainingInstances = new ConcurrentHashMap<>();
+    instances.forEach((k) -> remainingInstances.put(k, true));
+
+    final Properties props = ConnectionStringHelper.getDefaultProperties();
+    DriverHelper.setConnectTimeout(props, 5, TimeUnit.SECONDS);
+    DriverHelper.setSocketTimeout(props, 5, TimeUnit.SECONDS);
+
+    for (final String id : instances) {
+      executorService.submit(
+          () -> {
+            while (true) {
+              TestInstanceInfo instanceInfo =
+                  TestEnvironment.getCurrent().getInfo().getDatabaseInfo().getInstance(id);
+              try (final Connection conn =
+                  DriverManager.getConnection(
+                      ConnectionStringHelper.getUrl(
+                          instanceInfo.getEndpoint(),
+                          instanceInfo.getEndpointPort(),
+                          TestEnvironment.getCurrent()
+                              .getInfo()
+                              .getDatabaseInfo()
+                              .getDefaultDbName()),
+                      props)) {
+                remainingInstances.remove(id);
+                break;
+              } catch (final SQLException ex) {
+                ex.printStackTrace();
+                // Continue waiting until instance is up.
+              } catch (final Exception ex) {
+                System.out.println("Exception: " + ex);
+                break;
+              }
+              TimeUnit.MILLISECONDS.sleep(1000);
+            }
+            return null;
+          });
+    }
+    executorService.shutdown();
+    executorService.awaitTermination(5, TimeUnit.MINUTES);
+
+    if (finalCheck) {
+      assertTrue(
+          remainingInstances.isEmpty(),
+          "The following instances are still down: \n"
+              + String.join("\n", remainingInstances.keySet()));
+    }
+  }
+
+  public void failoverClusterAndWaitUntilWriterChanged(String clusterWriterId)
+      throws InterruptedException {
+    failoverClusterAndWaitUntilWriterChanged(
+        TestEnvironment.getCurrent().getInfo().getAuroraClusterName(), clusterWriterId);
+  }
+
+  public void failoverClusterAndWaitUntilWriterChanged(String clusterId, String clusterWriterId)
```

Comment:
There is a fix that will need to be ported over here where we need to wait for the writer IP to change. Can be done in a future PR if needed, as long as we don't forget